File: o11nplugin-contrail-vro-core/src/main/java/net/juniper/contrail/vro/ContrailPluginFactory.java
/*
 * Copyright (c) 2018 Juniper Networks, Inc. All rights reserved.
 */

package net.juniper.contrail.vro;

import java.util.*;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import ch.dunes.vso.sdk.api.HasChildrenResult;
import ch.dunes.vso.sdk.api.IPluginFactory;
import ch.dunes.vso.sdk.api.IPluginNotificationHandler;
import ch.dunes.vso.sdk.api.PluginExecutionException;
import ch.dunes.vso.sdk.api.QueryResult;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import static org.springframework.beans.factory.config.ConfigurableBeanFactory.*;

@Component
@Scope(SCOPE_PROTOTYPE)
public class ContrailPluginFactory implements IPluginFactory {

    private static final Logger log = LoggerFactory.getLogger(ContrailPluginFactory.class);

    private String pluginName;
    private String sessionId;
    private String username;
    private String password;
    private IPluginNotificationHandler pluginNotificationHandler;

    
    public ContrailPluginFactory(String pluginName, String sessionId, String username, String password, IPluginNotificationHandler pluginNotificationHandler) {
        log.debug("new ContrailPluginFactory() --> username: " + username);
        this.pluginName = pluginName;
        this.sessionId = sessionId;
        this.username = username;
        this.password = password;
        this.pluginNotificationHandler = pluginNotificationHandler;

    }

    @Override
    public void executePluginCommand(String cmd) throws PluginExecutionException {
        log.debug("executePluginCommand() --> cmd: " + cmd);
    }

    @Override
    public Object find(String type, String id) {
        return null;
    }

    @Override
    public QueryResult findAll(String type, String query) {
        log.debug("findAll() --> type: " + type + ", query: " + query);

        return new QueryResult();
    }

    @Override
    public HasChildrenResult hasChildrenInRelation(String parentType, String parentId, String relationName) {
        log.debug("hasChildrenInRelation() --> parentType: " + parentType + ", parentId: " + parentId + ", relationName: " + relationName);

        return HasChildrenResult.Unknown;
    }

    @Override
    public List<?> findRelation(String parentType, String parentId, String relationName) {
        log.debug("findRelation() --> parentType: " + parentType + ", parentId: " + parentId + ", relationName: " + relationName);

        
        return null;
    }

    @Override
    public void invalidate(String type, String id) {
        log.debug("invalidate() --> type: " + type + ", id: " + id);
    }

    @Override
    public void invalidateAll() {
        log.debug("invalidateAll()");
    }

    public String getUsername() {
        return username;
    }

    public IPluginNotificationHandler getPluginNotificationHandler() {
        return pluginNotificationHandler;
    }
}

File: o11nplugin-contrail-vro-core/src/main/kotlin/net/juniper/contrail/vro/ContrailPluginFactory.kt
/*
 * Copyright (c) 2018 Juniper Networks, Inc. All rights reserved.
 */

package net.juniper.contrail.vro

import net.juniper.contrail.vro.config.ConnectionManager
import org.slf4j.LoggerFactory
import ch.dunes.vso.sdk.api.HasChildrenResult
import ch.dunes.vso.sdk.api.IPluginFactory
import ch.dunes.vso.sdk.api.IPluginNotificationHandler
import ch.dunes.vso.sdk.api.PluginExecutionException
import ch.dunes.vso.sdk.api.QueryResult
import net.juniper.contrail.vro.config.ConnectionRepository
import net.juniper.contrail.vro.config.CONNECTION
import net.juniper.contrail.vro.config.ROOT
import net.juniper.contrail.vro.config.ROOT_HAS_CONNECTIONS
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component
import org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_PROTOTYPE

@Component
@Scope(SCOPE_PROTOTYPE)
class ContrailPluginFactory(
    val connections: ConnectionManager,
    val connectionRepository: ConnectionRepository,
    val pluginNotificationHandler: IPluginNotificationHandler) : IPluginFactory {

    private val log = LoggerFactory.getLogger(ContrailPluginFactory::class.java)

    init {
        log.debug("Created new ContrailPluginFactory")
    }

    @Throws(PluginExecutionException::class)
    override fun executePluginCommand(cmd: String) {
        log.debug("executePluginCommand(cmd='{}')", cmd)
    }

    override fun find(type: String, id: String?): Any? {
        log.debug("find(type={}, id={})", type, id)

        if (type == CONNECTION && id != null)
            return connectionRepository.getConnection(id)

        return null
    }

    private fun <T> List<T>.toResult(): QueryResult {
        val result = QueryResult()
        forEach(result::addElement)
        result.totalCount = size.toLong()
        return result
    }

    override fun findAll(type: String, query: String): QueryResult {
        log.debug("findAll(type={}, query='{}')", type, query)

        if (type == CONNECTION) {
            val connections = connectionRepository.findConnections(query)
            log.debug("Found {} connections", connections.size)
            return connections.toResult()
        }

        return QueryResult()
    }

    override fun hasChildrenInRelation(parentType: String, parentId: String?, relationName: String): HasChildrenResult {
        log.debug("hasChildrenInRelation(parentType={}, parentId={}, relationName={})", parentType, parentId, relationName)

        return HasChildrenResult.Unknown
    }

    override fun findRelation(parentType: String, parentId: String?, relationName: String?): List<*>? {
        log.debug("findRelation() --> parentType: $parentType, parentId: $parentId, relationName: $relationName")

        if (parentType == ROOT && relationName == ROOT_HAS_CONNECTIONS)
            return connectionRepository.connections

        return null
    }

    override fun invalidate(type: String?, id: String) {
        log.debug("invalidate(type={}, id={})", type, id)
    }

    override fun invalidateAll() {
        log.debug("invalidateAll()")
    }
}
